File: app/src/main/java/com/njamb/geodrink/Adapters/FriendListAdapter.java
package com.njamb.geodrink.Adapters;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.njamb.geodrink.ViewHolders.FriendViewHolder;
import com.njamb.geodrink.Models.User;
import com.njamb.geodrink.R;

import java.util.ArrayList;

public class FriendListAdapter extends RecyclerView.Adapter<FriendViewHolder> {
    private Context mContext;
    private ArrayList<User> mFriends = new ArrayList<>();
    private RecyclerView mRecyclerView;

    private DatabaseReference mUsersDatabase;

    public FriendListAdapter(final Context context, RecyclerView rv, User u) {
        mContext = context;
        mRecyclerView = rv;

        mUsersDatabase = FirebaseDatabase.getInstance().getReference("users");

        for (String key : u.friends.keySet()) {
            // Get friend info
            mUsersDatabase.child(key).addListenerForSingleValueEvent(new ValueEventListener() {
                @Override
                public void onDataChange(DataSnapshot dataSnapshot) {
                    mFriends.add(dataSnapshot.getValue(User.class));
                    notifyItemInserted(mFriends.size()-1);
                }

                @Override
                public void onCancelled(DatabaseError databaseError) {}
            });
        }
    }

    @Override
    public FriendViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(mContext);
        View view = inflater.inflate(R.layout.list_item_friends, parent, false);
        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                int index = mRecyclerView.getChildLayoutPosition(v);
                Toast.makeText(mContext,
                        String.format("%s says hi", mFriends.get(index).username),
                        Toast.LENGTH_SHORT).show();
            }
        });
        return new FriendViewHolder(view);
    }

    @Override
    public void onBindViewHolder(FriendViewHolder holder, int position) {
        User friend = mFriends.get(position);

        holder.tvUsername.setText(friend.username);
        holder.tvFullName.setText(friend.fullName);
        Glide.with(mContext)
                .load(friend.profileUrl)
                .apply(RequestOptions.errorOf(R.mipmap.geodrink_blue_logo))
                .apply(RequestOptions.circleCropTransform())
                .into(holder.ivProfileImg);
        holder.tvEmail.setText(friend.email);
    }

    @Override
    public int getItemCount() {
        return mFriends.size();
    }
}

File: app/src/main/java/com/njamb/geodrink/adapters/FriendListAdapter.kt
package com.njamb.geodrink.adapters

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast

import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.njamb.geodrink.R
import com.njamb.geodrink.models.User
import com.njamb.geodrink.view_holders.FriendViewHolder

import java.util.ArrayList

class FriendListAdapter(private val mContext: Context, private val mRecyclerView: RecyclerView, u: User) : RecyclerView.Adapter<FriendViewHolder>() {
    private val mFriends = ArrayList<User>()

    private val mUsersDatabase: DatabaseReference

    init {

        mUsersDatabase = FirebaseDatabase.getInstance().getReference("users")

        for (key in u.friends.keys) {
            // Get friend info
            mUsersDatabase.child(key).addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    mFriends.add(dataSnapshot.getValue(User::class.java))
                    notifyItemInserted(mFriends.size - 1)
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FriendViewHolder {
        val inflater = LayoutInflater.from(mContext)
        val view = inflater.inflate(R.layout.list_item_friends, parent, false)
        view.setOnClickListener { v ->
            val index = mRecyclerView.getChildLayoutPosition(v)
            // TODO: open profile for that user?
            Toast.makeText(mContext,
                           String.format("%s says hi", mFriends[index].username),
                           Toast.LENGTH_SHORT).show()
        }
        return FriendViewHolder(view)
    }

    override fun onBindViewHolder(holder: FriendViewHolder, position: Int) {
        val friend = mFriends[position]

        holder.tvUsername.text = friend.username
        holder.tvFullName.text = friend.fullName
        Glide.with(mContext)
                .load(friend.profileUrl)
                .apply(RequestOptions.errorOf(R.mipmap.geodrink_blue_logo))
                .apply(RequestOptions.circleCropTransform())
                .into(holder.ivProfileImg)
        holder.tvEmail.text = friend.email
    }

    override fun getItemCount(): Int {
        return mFriends.size
    }
}
